scikeras/utils/transformers.py

```diff
@@ -0,0 +1,179 @@
+import numpy as np
+import tensorflow as tf
+
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import FunctionTransformer, OneHotEncoder, OrdinalEncoder
+from sklearn.utils.multiclass import type_of_target
+from tensorflow.python.keras.losses import is_categorical_crossentropy
+
+
+class TargetReshaper(BaseEstimator, TransformerMixin):
+    def fit(self, X):
+        self.ndim_ = X.ndim
+        return self
+
+    def transform(self, X):
+        if self.ndim_ == 1:
```

Comment:
`self.ndim_` is set in `fit` because it is required by `inverse_transform` (if the input was 2D, we want the output to be 2D). So we need `fit`. In `transform`, we could just ignore the attribute set in `fit` and check `X` again, but I don't know that that's any better. I'd actually rather have a `ValueError` in `transform` if `X.ndim != self.ndim_` (which we had before but removed).

Reply:
In other words:
```python3
class TargetReshaper(BaseEstimator, TransformerMixin):
    def fit(self, X):
        self.ndim_ = X.ndim
        return self

    def transform(self, X):
        if X.ndim != self.ndim_:
             raise ValueError(...)
        if self.ndim_ == 1:
            return X.reshape(-1, 1)
        return X

    def inverse_transform(self, X):
        if self.ndim_ == 1 and X.ndim == 2:
            return np.squeeze(X, axis=1)
        return X
```

Reply:
Or maybe we should just go back to https://github.com/adriangb/scikeras/blob/d5532ea694d3e1096242884c97b01b9d0d760569/scikeras/utils/transformers.py#L13

It's the same as what we have now, except that it validates that the inputs/outputs are the expected shape and raises ValueErrors if they're not.

Reply:
> In transform, we could just ignore the attribute set in fit and check X again, but I don't know that that's any better. 

Yes. That's the implementation that I'd like to see. Then it'd be a stateless transform. That would enable this use case:

``` python
from sklearn.datasets import make_classification

class TargetReshaper(BaseEstimator, TransformerMixin):
    def fit(self, X):
        self.ndim_ = X.ndim
        return self

    @staticmethod
    def transform(X):
        if X.ndim == 1:
            return X.reshape(-1, 1)
        return X

    def inverse_transform(self, X):
        if not hasattr(self, "ndim_"):
            raise NotFittedError(...)
        if self.ndim_ == 1 and X.ndim == 2:
            return np.squeeze(X, axis=1)
        return X

X, y = make_classification()
y2d = TargetReshaper.transform(y)
```

This implementation raises an error only when necessary, and it's also more informative.

Reply:
Sure, sounds good, I'll push that in a sec.

Reply:
Implemented. Resolving this thread.